src/fastapibootcamp/handlers/external.py

```diff
@@ -317,3 +330,135 @@ async def post_log_demo(
+# an external service is down, or some bug in your code causes an exception.
+# By default, FastAPI returns a 500 Internal Server Error response when an
+# uncaught exception occurs. Safir provides the ability to report these
+# uncaught errors to a Slack channel through an incoming webhook.
+#
+# To hook up this Slack reporting, you need to take the following steps
+# 1. Add a slack_webhook_url field to the configuration (config.py)
+# 2. In main.py, configure the SlackRouteErrorHandler middleware
+# 3. Create custom SlackException subclasses for your application (usually
+#    in exceptions.py)
+# 4. Raise these exceptions in your application code
+# 5. Optionally, add SlackRouteErrorHandler to the APIRouter in your
+#    handlers/endpoints module to automatically report any uncaught exception
+#    to Slack.
+#
+# If your error is caused by user input, you typically don't want to report
+# is to Slack, but instead to the user. In the astroplan application we'll
+# explore the ClientRequestError exception for this purpose in
+# handlers/astroplan/endpoints.py's get_observer function.
+#
+# Try it out:
+#   http post :8000/fastapi-bootcamp/error-demo custom_error:=true
+
+
+class ErrorRequestModel(BaseModel):
+    """Request model for the error POST endpoint."""
+
+    custom_error: bool = Field(
+        True, title="If true, raise the custom SlackException."
+    )
+
+
+@external_router.post(
+    "/error-demo", summary="Raise an internal service exception."
+)
+async def post_error_demo(data: ErrorRequestModel) -> JSONResponse:
+    """Raise an internal service exception."""
+    if data.custom_error:
+        raise DemoInternalError(
+            "A custom error occurred.", custom_data="Hello error!"
+        )
+
+    raise RuntimeError("A generic error occurred.")
+
+
+# =============================================================================
+# Lesson 6: Custom dependencies
+#
+# FastAPI dependencies are a way to add reusable code to your path operation
+# that's aware of the current request. Safir provides several dependencies,
+# see https://safir.lsst.io/api.html#module-safir.dependencies.arq etc.
+#
+# - arq_dependency provides a client to the Arq distributed job queue
+# - db_session_dependency provides a SQLAlchemy session
+# - auth_delegated_token_dependency provides a delegated token
+# - auth_dependency provides info about the current user
+# - auth_logger_dependency provides a logger with user info bound
+# - http_client_dependency provides an HTTPX async client
+# - logger_dependency provides a structlog logger (see Lesson 4)
+#
+# Besides these, you can create your own dependencies. In the astroplan
+# application we'll explore the request context dependency pattern.
+#
+# There are two types of dependencies you'll develop: functional dependencies
+# and class-based dependencies. These class-based dependencies can hold
+# persistent state that's reused across multiple requests.
```

Comment:
What about "singleton dependencies"? That seems nicely parallel to functional dependencies, since the dependency is a singleton instead of a function.

Reply:
I thought I covered that with my stateful example? I'm
Not quite sure what you mean.

Reply:
I mean for terminology: instead of "class-based dependencies", how about "singleton dependencies"? Although maybe singleton is obscure terminology. I spend too much time thinking about type theory so I'm a bad person to ask.